File: python/2015/04_the_ideal_stocking_stuffer/aoc201504.py
"""AoC 4, 2015: The Ideal Stocking Stuffer"""

# Standard library imports
import hashlib
import itertools
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input"""
    return puzzle_input


def part1(data):
    """Solve part 1"""
    return first_md5(secret=data, prefix="0" * 5)


def part2(data):
    """Solve part 2"""
    return first_md5(secret=data, prefix="0" * 6)


def first_md5(secret, prefix):
    """Find the first advent coin with the given prefix

    ## Example:

    >>> first_md5("aoc", "000")
    425
    """
    for num in itertools.count(1):
        md5 = hashlib.md5()
        md5.update(f"{secret}{num}".encode("utf-8"))

        if md5.hexdigest().startswith(prefix):
            return num


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2015/04_the_ideal_stocking_stuffer/aoc201504.ex
defmodule AOC2015.Day04 do
  @moduledoc """
  Advent of Code 2015, day 4: The Ideal Stocking Stuffer
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n") |> hd()
  end

  def part1(input, prefix \\ "00000") do
    input |> find_md5_with_prefix(prefix)
  end

  def part2(input, prefix \\ "000000") do
    input |> find_md5_with_prefix(prefix)
  end

  @doc """
  Find the first MD5 hash of the form <secret><number> whose hex representation starts with prefix

  ## Examples:

      iex> find_md5_with_prefix("elixir", "A0C")
      2947

      iex> :crypto.hash(:md5, "elixir2947") |> Base.encode16() |> String.slice(0..2)
      "A0C"
  """
  def find_md5_with_prefix(secret, prefix) do
    Stream.cycle([1])
    |> Enum.reduce_while(0, fn step, number ->
      case :crypto.hash(:md5, "#{secret}#{number + step}")
           |> Base.encode16()
           |> String.starts_with?(prefix) do
        true -> {:halt, number + step}
        false -> {:cont, number + step}
      end
    end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
